src/FTPObject.jl

```diff
@@ -132,168 +132,117 @@ end
 """
     upload(
         ftp::FTP,
-        local_name::AbstractString;
+        local_path_io::IO,
+        remote_path::AbstractString;
+        ftp_options=ftp.ctxt,
         mode::FTP_MODE=binary_mode,
-    )
-
-Upload the file "local_name" to the FTP server and save as "local_name".
-"""
-function upload(
-    ftp::FTP,
-    local_name::AbstractString;
-    mode::FTP_MODE=binary_mode,
-    verbose=nothing,
-)
-    _dep_verbose_kw(verbose, typeof(ftp), :upload)
-    return upload(ftp, local_name, local_name; mode=mode, verbose=verbose)
-end
-
-"""
-    upload(
-        ftp::FTP,
-        local_name::AbstractString,
-        remote_name::AbstractString;
-        mode::FTP_MODE=binary_mode,
-    )
+        verbose=nothing
+) -> Response
 
-Upload the file "local_name" to the FTP server and save as "remote_name".
-"""
-function upload(
-    ftp::FTP,
-    local_name::AbstractString,
-    remote_name::AbstractString;
-    mode::FTP_MODE=binary_mode,
-    verbose=nothing,
-)
-    _dep_verbose_kw(verbose, typeof(ftp), :upload)
+Upload IO object "local_path_io" to the FTP server and save as "remote_path".
 
-    open(local_name) do local_file
-        return upload(ftp, local_file, remote_name; mode=mode, verbose=verbose)
-    end
-end
+# Arguments
+- `ftp::FTP`: The FTP to deliver to. See FTPClient.FTP for details.
+- `local_path_io::IO`: The IO object that we want to deliver.
+- `remote_path::AbstractString`: The path that we want to deliver to.
 
-"""
-    upload(
-        ftp::FTP,
-        local_file::IO,
-        remote_name::AbstractString;
-        mode::FTP_MODE=binary_mode,
-    )
+# Keywords
+- `ftp_options=ftp.ctxt`: FTP Options
+- `mode::FTP_MODE=binary_mode`: Set the ftp mode.
+- `verbose=nothing`: Set the verbosity
 
-Upload IO object "local_file" to the FTP server and save as "remote_name".
+# Returns
+`FTPResponse`: Returns the ftp response object
 """
 function upload(
     ftp::FTP,
-    local_file::IO,
-    remote_name::AbstractString;
+    local_path_io::IO,
+    remote_path::AbstractString;
+    ftp_options=ftp.ctxt,
     mode::FTP_MODE=binary_mode,
     verbose=nothing,
 )
     _dep_verbose_kw(verbose, typeof(ftp), :upload)
 
+    resp = nothing
+
     try
-        ftp_put(ftp.ctxt, remote_name, local_file; mode=mode, verbose=verbose)
-    catch err
-        if isa(err, FTPClientError)
-            err.msg = "Failed to upload $remote_name."
+        resp = ftp_put(ftp_options, remote_path, local_path_io; mode=mode, verbose=verbose)
+    catch e
+        if isa(e, FTPClientError)
+            e.msg = "Failed to upload $remote_path"
```

Comment:
Should we append the original message?

Reply:
I believe this this case e.msg is empty and the actual error that exists in e is still shown, but e.msg is just some additional info that gets added to the output. I can check to make sure though.

Reply:
Hm yeah looks like it'll just spit out the lib_curl_error and also err.msg if it is defined. https://github.com/invenia/FTPClient.jl/blob/master/src/FTPClient.jl#L17

Reply:
Ok, we should probably only do this when `isempty(e.msg)` then